File: src/main/java/com/zhihao/admin/common/util/TreeUtils.java
package com.zhihao.admin.common.util;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Function;

public final class TreeUtils {

    private TreeUtils() {
    }

    public static <T> List<T> build(List<T> nodes, Function<T, Long> idGetter, Function<T, Long> parentIdGetter,
                                    Function<T, Integer> sortGetter, BiConsumer<T, T> childAppender, Long rootParentId) {
        Map<Long, T> nodeMap = new LinkedHashMap<>();
        nodes.stream()
                .sorted(Comparator.comparing(sortGetter, Comparator.nullsLast(Integer::compareTo)))
                .forEach(node -> nodeMap.put(idGetter.apply(node), node));

        List<T> roots = new ArrayList<>();
        for (T node : nodeMap.values()) {
            Long parentId = parentIdGetter.apply(node);
            T parent = nodeMap.get(parentId);
            if (parent == null || rootParentId.equals(parentId)) {
                roots.add(node);
            } else {
                childAppender.accept(parent, node);
            }
        }
        return roots;
    }
}

File: server/src/common/tree-utils.ts
export function buildTree<T extends { children: T[] }>(
  nodes: T[],
  idGetter: (node: T) => number,
  parentIdGetter: (node: T) => number,
  sortGetter: (node: T) => number | null | undefined,
  childAppender: (parent: T, child: T) => void,
  rootParentId: number,
): T[] {
  const sorted = [...nodes].sort((a, b) => {
    const aSort = sortGetter(a) ?? Number.MAX_SAFE_INTEGER
    const bSort = sortGetter(b) ?? Number.MAX_SAFE_INTEGER
    return aSort - bSort
  })

  const nodeMap = new Map<number, T>()
  for (const node of sorted) {
    nodeMap.set(idGetter(node), node)
  }

  const roots: T[] = []
  for (const node of nodeMap.values()) {
    const parentId = parentIdGetter(node)
    const parent = nodeMap.get(parentId)
    if (!parent || parentId === rootParentId) {
      roots.push(node)
    } else {
      childAppender(parent, node)
    }
  }

  return roots
}
